### PR TITLE
const-correct output methods

### PR DESCRIPTION
**Pull request by david-eckardt-sociomantic**
Change the types of output buffers to `const`, all the way from the `RequestOnConn` API methods down to the socket output function.

### CODE REVIEW

**Review by mihails-strasuns-sociomantic — APPROVED** [review]
All `cast(Const!(void*))` bits looks somewhat scary but I couldn't find anything practically wrong.

File: src/swarm/neo/protocol/socket/uio_const.d [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+{
+    Const!(msghdr) m = {
+            msg_iov:    cast(Const!(iovec*))data.ptr,
+            msg_iovlen: data.length
+    };
+    return sendmsg(socket_fd, &m, flags);
+}
+
+// Ensure binary compatibility of `iovec_const` with `iovec`.
+unittest
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/protocol/socket/uio_const.d:144]: Would be cool to have a template in ocean: `BinaryLayoutIdentical`, or some such. (Future suggestion.)

File: src/swarm/neo/protocol/socket/uio_const.d [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+module swarm.neo.protocol.socket.uio_const;
+
+import core.sys.posix.sys.types: ssize_t;
+
+import core.sys.posix.sys.socket;
+
+import ocean.transition;
+
+/// D-`const`-correct output version of `iovec`.
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/protocol/socket/uio_const.d]: Would be nice to document why this is declared and how it is used.

File: src/swarm/neo/protocol/socket/uio_const.d [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    Returns:
+        0 on success or -1 on failure, in which case `errno` is set accordingly.
+
+*******************************************************************************/
+
+ssize_t sendv ( int socket_fd, in iovec_const[] data, int flags = 0 )
+{
+    Const!(msghdr) m = {
+            msg_iov:    cast(Const!(iovec*))data.ptr,
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/protocol/socket/uio_const.d:137]: Why is this cast needed? Afaics, `msghdr.msg_iov` is declared as a simple `iovec*` (non-const).

Comment (david-eckardt-sociomantic) [reply src/swarm/neo/protocol/socket/uio_const.d:137]: Hm, good question.

Comment (david-eckardt-sociomantic) [reply src/swarm/neo/protocol/socket/uio_const.d:137]: `typeof(msghdr.msg_iov)` is `iovec*` but `typeof(data.ptr)` is `iovec_const*`.

File: src/swarm/neo/protocol/socket/uio_const.d (hunk 1 above)
Comment (david-eckardt-sociomantic) [reply src/swarm/neo/protocol/socket/uio_const.d:144]: All the `sizeof`, `alignof` and `tupleof.length` checks could go in there but not easily the field type identity and equivalence checks: It verifies that the types of all fields are identical to the original struct except those that were changed to equivalent types (identical but `const` in this case). I don't see a simple way how an automatic checker could know which fields should be identical and which criterion of equivalence applies to those that should be different.

**Comment by david-eckardt-sociomantic** [comment]
Updated with a description of the purpose of `iovec_const` for this particular case and `uio` vector I/O in general.

File: src/swarm/neo/protocol/socket/uio_const.d (hunk 3 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/protocol/socket/uio_const.d:137]: Ah, of course :D

File: src/swarm/neo/protocol/socket/uio_const.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/protocol/socket/uio_const.d:144]: Well, I imagined using `Unqual` on everything to compare types.

**Comment by gavin-norman-sociomantic** [comment]
LGTM. I can't remember... did we discuss whether this should be applied to v2.6.x?

**Comment by david-eckardt-sociomantic** [comment]
Moved to the private repository to be merged into v2.6.x.

**Pull request closed without merging**